<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 3.0//EN">
<!--Converted with LaTeX2HTML 96.1 (Feb 5, 1996) by Nikos Drakos (nikos@cbl.leeds.ac.uk), CBLU, University of Leeds -->
<HTML>
<HEAD>
<TITLE>8.2 Exceptions</TITLE>
<META NAME="description" CONTENT="8.2 Exceptions">
<META NAME="keywords" CONTENT="tut">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">
<LINK REL=STYLESHEET HREF="tut.css">
</HEAD>
<BODY LANG="EN">
 <A NAME="tex2html799" HREF="node49.html"><IMG WIDTH=37 HEIGHT=24 ALIGN=BOTTOM ALT="next" SRC="../icons/next_motif.gif" BORDER=0></A> <A NAME="tex2html797" HREF="node46.html"><IMG WIDTH=26 HEIGHT=24 ALIGN=BOTTOM ALT="up" SRC="../icons/up_motif.gif" BORDER=0></A> <A NAME="tex2html791" HREF="node47.html"><IMG WIDTH=63 HEIGHT=24 ALIGN=BOTTOM ALT="previous" SRC="../icons/previous_motif.gif" BORDER=0></A> <A NAME="tex2html801" HREF="node1.html"><IMG WIDTH=65 HEIGHT=24 ALIGN=BOTTOM ALT="contents" SRC="../icons/contents_motif.gif" BORDER=0></A>  <BR>
<B> Next:</B> <A NAME="tex2html800" HREF="node49.html">8.3 Handling Exceptions</A>
<B>Up:</B> <A NAME="tex2html798" HREF="node46.html">8 Errors and Exceptions</A>
<B> Previous:</B> <A NAME="tex2html792" HREF="node47.html">8.1 Syntax Errors</A>
<BR><HR><P>
<H1><A NAME="SECTION00920000000000000000">8.2 Exceptions</A></H1>
<P>
Even if a statement or expression is syntactically correct, it may
cause an error when an attempt is made to execute it.
Errors detected during execution are called <EM>exceptions</EM> and are
not unconditionally fatal: you will soon learn how to handle them in
Python programs.  Most exceptions are not handled by programs,
however, and result in error messages as shown here:
<P>
<SMALL CLASS=SMALL><PRE>&gt;&gt;&gt; 10 * (1/0)
Traceback (innermost last):
  File &quot;&lt;stdin&gt;&quot;, line 1
ZeroDivisionError: integer division or modulo
&gt;&gt;&gt; 4 + spam*3
Traceback (innermost last):
  File &quot;&lt;stdin&gt;&quot;, line 1
NameError: spam
&gt;&gt;&gt; '2' + 2
Traceback (innermost last):
  File &quot;&lt;stdin&gt;&quot;, line 1
TypeError: illegal argument type for built-in operation
&gt;&gt;&gt;</PRE></SMALL>
The last line of the error message indicates what happened.
Exceptions come in different types, and the type is printed as part of
the message: the types in the example are
<TT>ZeroDivisionError</TT>,
<TT>NameError</TT>
and
<TT>TypeError</TT>.
The string printed as the exception type is the name of the built-in
name for the exception that occurred.  This is true for all built-in
exceptions, but need not be true for user-defined exceptions (although
it is a useful convention).
Standard exception names are built-in identifiers (not reserved
keywords).
<P>
The rest of the line is a detail whose interpretation depends on the
exception type; its meaning is dependent on the exception type.
<P>
The preceding part of the error message shows the context where the
exception happened, in the form of a stack backtrace.
In general it contains a stack backtrace listing source lines; however,
it will not display lines read from standard input.
<P>
The Python Library Reference Manual lists the built-in exceptions and
their meanings.
<P>
<BR> <HR>
<P><ADDRESS>
guido@cnri.reston.va.us
</ADDRESS>
</BODY>
</HTML>
